build.assets/tooling/cmd/resource-ref-generator/reference/reference.go

```diff
@@ -0,0 +1,332 @@
+//   - is a struct type
+//   - has fields with the required types
+//   - does not belong to the list of excluded resources
+func shouldProcess(d resource.DeclarationInfo, requiredTypes, excludedResources []TypeInfo) bool {
+	// We expect the declaration to be a type declaration with one spec.
+	gendecl, ok := d.Decl.(*ast.GenDecl)
+	if !ok {
+		return false
+	}
+
+	if len(gendecl.Specs) != 1 {
+		return false
+	}
+
+	t, ok := gendecl.Specs[0].(*ast.TypeSpec)
+	if !ok {
+		return false
+	}
+
+	if t == nil {
+		return false
+	}
+
+	// If the declaration type is not a struct, we can't process it as a
+	// root resource entry.
+	str, ok := t.Type.(*ast.StructType)
+	if !ok {
+		return false
+	}
+
+	// If the configured excluded resources include this type declaration,
+	// don't process it.
+	for _, r := range excludedResources {
+		if t.Name.Name == r.Name && d.PackageName == r.Package {
+			return false
+		}
+	}
+	// Use only the final segment of each desired package path
+	// in the comparison, since that is what is preserved in the
+	// AST.
+	finalTypes := make([]TypeInfo, len(requiredTypes))
+	for i, t := range requiredTypes {
+		segs := strings.Split(t.Package, "/")
+		finalTypes[i] = TypeInfo{
+			Package: segs[len(segs)-1],
+			Name:    t.Name,
+		}
+	}
+
+	// Compare the types of fields in the struct type with the required
+	// fields types. Only one required field type must be present.
+	var m bool
+	for _, fld := range str.Fields.List {
+		if len(fld.Names) != 1 {
+			continue
+		}
+
+		// Identify a package for the field type so we can check it
+		// against the required field list. Begin by assuming the field
+		// comes from the same package as the outer struct type, then
+		// assign a package name depending on the expression used to
+		// declare the field type.
+		gopkg := d.PackageName
+		var fldname string
+		switch t := fld.Type.(type) {
+		case *ast.SelectorExpr:
+			// If the type of the field is an *ast.SelectorExpr,
+			// it's of the form <package>.<type name>.
+			g, ok := t.X.(*ast.Ident)
+			if ok {
+				gopkg = g.Name
+			}
+			fldname = t.Sel.Name
+
+		// There's no package, so only assign a name.
+		case *ast.Ident:
+			fldname = t.Name
+		}
+
+		for _, ti := range finalTypes {
+			if gopkg == ti.Package && fldname == ti.Name {
+				m = true
+				break
+			}
+		}
+	}
+
+	return m
+}
+
+type GenerationError struct {
+	messages []error
+}
+
+func (g GenerationError) Error() string {
+	// Begin with a newline to format the first list item below the outer
+	// error.
+	final := "\n"
+	for _, e := range g.messages {
+		final += fmt.Sprintf("- %v\n", e)
+	}
+	return final
+}
+
+// Generate uses the provided user-facing configuration to write the resource
+// reference to fs.
+func Generate(srcFS, destFS afero.Fs, conf GeneratorConfig) error {
+	sourceData, err := resource.NewSourceData(srcFS, conf.SourcePath)
+	if err != nil {
+		return fmt.Errorf("can't load Go source files: %v", err)
```

Comment:
If we use %w, then the error is properly wrapped.

```suggestion
		return fmt.Errorf("loading Go source files: %w", err)
```